mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java

```diff
@@ -0,0 +1,187 @@
+import com.mongodb.hadoop.util.MongoConfigUtil;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.Job;
+import org.bson.BSONObject;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSource} that reads data from MongoDB and converts each document into 
+ * a {@link StructuredRecord} with the help of the specified Schema.
+ */
+@Plugin(type = "batchsource")
+@Name("MongoDB")
+@Description("MongoDB Batch Source will read documents from MongoDB and convert each document " +
+  "into a StructuredRecord with the help of the specified Schema. " +
+  "Optionally, the user can specify input query, input fields, and splitter class.")
+public class MongoDBBatchSource extends BatchSource<Object, BSONObject, StructuredRecord> {
+
+  private static final Gson GSON = new Gson();
+  private final MongoDBConfig config;
+
+  public MongoDBBatchSource(MongoDBConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = context.getHadoopJob();
+    Configuration conf = job.getConfiguration();
+    MongoConfigUtil.setInputFormat(conf, MongoInputFormat.class);
+    MongoConfigUtil.setInputURI(conf, config.connectionString);
+    if (!Strings.isNullOrEmpty(config.inputQuery)) {
+      MongoConfigUtil.setQuery(conf, config.inputQuery);
+    }
+    if (!Strings.isNullOrEmpty(config.authConnectionString)) {
+      MongoConfigUtil.setAuthURI(conf, config.authConnectionString);
+    }
+    if (!Strings.isNullOrEmpty(config.inputFields)) {
+      MongoConfigUtil.setFields(conf, config.inputFields);
+    }
+    if (!Strings.isNullOrEmpty(config.splitterClass)) {
+      String className = String.format("%s.%s", StandaloneMongoSplitter.class.getPackage().getName(),
+                                       config.splitterClass);
+      Class<? extends MongoSplitter> klass = getClass().getClassLoader().loadClass(
+        className).asSubclass(MongoSplitter.class);
+      MongoConfigUtil.setSplitterClass(conf, klass);
+    }
+    Schema.parseJson(config.schema);
+    job.setInputFormatClass(MongoConfigUtil.getInputFormat(conf));
+  }
+
+  @Override
+  public void transform(KeyValue<Object, BSONObject> input, Emitter<StructuredRecord> emitter) throws Exception {
+    BSONObject bsonObject = input.getValue();
+    Schema schema = Schema.parseJson(config.schema);
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    for (Schema.Field field : schema.getFields()) {
+      builder.set(field.getName(), convertValue(bsonObject.get(field.getName()).toString(), field));
+    }
+    emitter.emit(builder.build());
+  }
+
+  private Object convertValue(String input, Schema.Field field) throws IOException {
+    switch (field.getSchema().getType()) {
+      case NULL:
+        return null;
+      case BOOLEAN:
+        return Boolean.valueOf(input);
```

Comment:
how come we need to convert everything to a string and then convert back? does BSONObject not return the correct type?